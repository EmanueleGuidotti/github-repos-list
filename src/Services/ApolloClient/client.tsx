import {ApolloClient, ApolloLink, InMemoryCache, HttpLink} from "@apollo/client";
import {setContext} from "@apollo/client/link/context";

const httpLink = new HttpLink({
    uri: "https://api.github.com/graphql",
});

const authLink = setContext((_, {headers}) => {
    const token = process.env.REACT_APP_API_KEY;
    return {
        headers: {
            ...headers,
            authorization: token ? `Bearer ${token}` : "",
        },
    };
});

export const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: authLink.concat(ApolloLink.from([httpLink])),
});
