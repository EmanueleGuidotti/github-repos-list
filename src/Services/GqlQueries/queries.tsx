import {gql} from "@apollo/client";

export const GET_REPOS = gql`
  query ($start: Int!, $end: String) {
    search(
      type: REPOSITORY
      query: """
      name: "react"
      """
      first: $start
      after: $end
    ) {
      pageInfo {
        startCursor
        hasNextPage
        endCursor
      }
      repos: edges {
        repo: node {
          ... on Repository {
            id
            name
            url
            forks {
              totalCount
            }
            stargazers {
              totalCount
            }
          }
        }
      }
    }
  }
`;

export const GET_REPOS_TOTAL = gql`
  query {
    search(
      type: REPOSITORY
      query: """
      name: "react"
      """
    ) {
      repositoryCount
    }
  }
`;
