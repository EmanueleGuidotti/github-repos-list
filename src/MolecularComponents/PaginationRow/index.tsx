import React from "react";
import {OnClickParent, ReposDataTotalResult} from "../../Interfaces/interfaces";
import {useQuery} from "@apollo/client";
import {GET_REPOS_TOTAL} from "../../Services/GqlQueries/queries";
import GenericButton from "../../AtomicComponents/GenericButton";

const result = 100;

const PaginationRow = ({onClickParent}: OnClickParent) => {
    const {data, loading, error} = useQuery<ReposDataTotalResult>(GET_REPOS_TOTAL);

    if (loading) {
        return <div>loading</div>;
    }

    if (error) {
        return <div>error fetching data</div>;
    }

    const onClickCursor = (): void => {
        onClickParent();
    };

    return (
        <div style={{display: "flex", flexDirection: "row"}}>
            {Array.from(Array(data?.search.repositoryCount), (e, i) => {
                if (i % result === 0 || i == 0)
                    return <GenericButton key={i} text={i} onClickCursor={() => onClickCursor()}></GenericButton>;
            })}
        </div>
    );
};

export default PaginationRow;
