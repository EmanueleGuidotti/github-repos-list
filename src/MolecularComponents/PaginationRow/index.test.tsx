import React from "react";
import renderer, {act} from "react-test-renderer";
import PaginationRow from "./index";
import TestRenderer from "react-test-renderer";
import {MockedProvider} from "@apollo/client/testing";
import {waitFor} from "@testing-library/react";
import {GET_REPOS_TOTAL} from "../../Services/GqlQueries/queries";
import {GraphQLError} from "graphql";

const SuccessMocks = [
    {
        request: {
            query: GET_REPOS_TOTAL,
        },
        result: {
            data: {
                "search":
                        {
                            "repositoryCount": 3378,
                            "__typename": "SearchResultItemConnection"
                        }
                }
            },
        },
];

const ErrorMocks = [
    {
        request: {
            query: GET_REPOS_TOTAL,
        },
        result: {
            errors: [new GraphQLError("Error!")],
        },
    },
];

describe("TableRowComponent", () => {
    it("render pagination loading state", () => {
        const component = TestRenderer.create(
            <MockedProvider mocks={SuccessMocks} addTypename={false}>
                <PaginationRow onClickParent={()=>{}} />
            </MockedProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("render pagination error state", async () => {
        const component = TestRenderer.create(
            <MockedProvider mocks={ErrorMocks} addTypename={false}>
                <PaginationRow onClickParent={()=>{}} />
            </MockedProvider>
        );
        await act(async () => {
            await waitFor(() => new Promise((resolve) => setTimeout(resolve, 0)));
        });
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("render pagination row component", async () => {
        const component = TestRenderer.create(
            <MockedProvider mocks={SuccessMocks} addTypename={false}>
                <PaginationRow onClickParent={() => {}} />
            </MockedProvider>
        );
        await act(async () => {
            await waitFor(() => new Promise((resolve) => setTimeout(resolve, 0)));
        });
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
