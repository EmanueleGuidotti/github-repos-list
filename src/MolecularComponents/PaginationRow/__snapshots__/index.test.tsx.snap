// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRowComponent render pagination error state 1`] = `
<div>
  error fetching data
</div>
`;

exports[`TableRowComponent render pagination loading state 1`] = `
<div>
  loading
</div>
`;

exports[`TableRowComponent render pagination row component 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    0
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    100
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    200
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    300
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    400
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    500
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    600
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    700
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    800
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    900
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1000
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1100
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1200
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1300
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1400
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1500
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1600
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1700
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1800
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    1900
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2000
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2100
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2200
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2300
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2400
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2500
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2600
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2700
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2800
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    2900
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    3000
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    3100
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    3200
  </button>
  <button
    onClick={[Function]}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": "1",
        "justifyContent": "center",
        "padding": "5",
      }
    }
  >
    3300
  </button>
</div>
`;
